emukit/bayesian_optimization/acquisitions/local_penalization.py

```diff
@@ -0,0 +1,98 @@
+from typing import Tuple
+
+import numpy as np
+from scipy.stats import norm
+
+from emukit.core.acquisition import Acquisition
+from emukit.core.interfaces import IModel
+
+
+class LocalPenalization(Acquisition):
+    """
+    Penalization function used in the local penalization method. It is the log of the following function:
+
+    .. math::
+        h(x) = \sum_i{\log\\left(\Phi\\left(\\frac{\|x - x_i\|_{2} - r_i}{s_i}\\right)\\right)}
+
+    where :math:`x_i` are the points already in the batch
+    """
+    def __init__(self, model: IModel):
+        """
+        :param model: Model
+        """
+
+        self.x_batch = None
+        self.radius = None
+        self.scale = None
+        self.model = model
+
+    @property
+    def has_gradients(self) -> bool:
+        return True
+
+    def update_batches(self, x_batch: np.ndarray, lipschitz_constant: float, f_min: float):
+        """
+        Updates the batches internally and pre-computes the parameters of the penalization function
+        """
+        self.x_batch = x_batch
+        if x_batch is not None:
+            self._compute_parameters(x_batch, lipschitz_constant, f_min)
+        else:
+            self.x_batch = None
+            self.radius = None
+            self.scale = None
+
+    def _compute_parameters(self, x_batch, lipschitz_constant, f_min):
+        """
+        Pre-computes the parameters of a penalization function
+        """
+        mean, variance = self.model.predict(x_batch)
+
+        variance = np.maximum(1e-16, variance)
```

Comment:
What is this number, `1e-16`? Either a comment or a good variable name would help understand that